pantheon-bundle/src/main/java/com/redhat/pantheon/servlet/sitemap/SiteMapServlet.java

```diff
@@ -0,0 +1,172 @@
+import com.redhat.pantheon.helper.PantheonConstants;
+import com.redhat.pantheon.model.api.Child;
+import com.redhat.pantheon.model.assembly.AssemblyVariant;
+import com.redhat.pantheon.model.document.DocumentMetadata;
+import com.redhat.pantheon.model.document.DocumentVariant;
+import com.redhat.pantheon.model.document.DocumentVersion;
+import com.redhat.pantheon.model.module.ModuleVariant;
+import org.apache.sling.api.SlingHttpServletRequest;
+import org.apache.sling.api.SlingHttpServletResponse;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.servlets.SlingAllMethodsServlet;
+import org.apache.sling.resource.filter.ResourceFilterStream;
+import org.apache.sling.servlets.annotations.SlingServletResourceTypes;
+import org.osgi.framework.Constants;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.servlet.Servlet;
+import javax.servlet.http.HttpServletResponse;
+import javax.xml.stream.XMLOutputFactory;
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.XMLStreamWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static com.redhat.pantheon.helper.PantheonConstants.*;
+
+@Component(
+        service = Servlet.class,
+        property = {
+                Constants.SERVICE_DESCRIPTION + "=Sitemap Servlet",
+                Constants.SERVICE_VENDOR + "=Red Hat Content Tooling team",
+                "sling.servlet.paths="+ "[\"/api/sitemap/module.sitemap.xml\", \"/api/sitemap/assembly.sitemap.xml\" ]"
+        }
+)
+@SlingServletResourceTypes(
+        resourceTypes = { SLING_SERVLET_DEFAULT },
+        methods = "GET",
+        extensions = SITEMAP_EXTENSION,
+        selectors = SITE_MAP)
+public class SiteMapServlet extends SlingAllMethodsServlet {
+    private final Logger log = LoggerFactory.getLogger(SiteMapServlet.class);
+
+    private static final String RESOURCE_ROOT = "/content/repositories";
+
+    private List<Resource> getAsset(Resource resource, String documentVariantResourceType, String documentVersionResourceType, List<Resource> assets) {
+
+        for (Resource r1 : resource.getChildren()) {
+            // Use Resource Filter Stream to limit memory consumption and path traversal
+            ResourceFilterStream rfs = r1.adaptTo(ResourceFilterStream.class);
+
+            assets.addAll(rfs.setBranchSelector("[sling:resourceType] == '" + documentVariantResourceType + "'")
+                    .setChildSelector("[released/sling:resourceType] == '" + documentVersionResourceType + "'")
+                    .stream()
+                    .collect(Collectors.toList()));
+            if (r1.hasChildren()) {
+                getAsset(r1, documentVariantResourceType, documentVersionResourceType, assets);
+            }
+        }
```

Comment:
Is there a more efficient way to gather all children? Even if you are adapting to a stream, which is a cool technique that I have never seen before, this is still a loop that performs a database query for every single node in the tree. That could be a very expensive operation.

Reply:
I am using Resource Filter Stream service because is way more efficient than a database query. 
https://sling.apache.org/documentation/bundles/resource-filter.html

I also have a branch that uses one sql2 query to get the released assets, which is very slower on dev/qa/stage. It reminds me of the search home page. I have to wait for a while to see the display.  I'll deploy both branch on ohc so that you can see the difference in performance gain using Resource Filter Stream.

Reply:
Ben, good call on the loop. It's removed.